<a href="https://colab.research.google.com/github/anishm91/HR_Analytics-Attrition_Prediction/blob/master/Attrition_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Getting the Data from Google Drive:

In [1]:
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

    100% |████████████████████████████████| 993kB 23.8MB/s 


In [0]:
link = 'https://drive.google.com/open?id=1LGzFPZ5AEVFRWjr2OKiZLNDzf4zwX5GV'

fluff, id = link.split('=')

downloaded = drive.CreateFile({'id':id}) 
downloaded.GetContentFile('HR.xlsx') 

#**Import Libraries**

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.options.display.max_columns = 99

#** Load Data**

In [0]:
df = pd.read_excel('HR.xlsx',sheet_name = 'WA_Fn-UseC_-HR-Employee-Attriti')

In [5]:
df.head(1)

Age Attrition BusinessTravel  DailyRate Department  DistanceFromHome  \
0   41       Yes  Travel_Rarely       1102      Sales                 1   

   Education EducationField  EmployeeCount  EmployeeNumber  \
0          2  Life Sciences              1               1   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   

           JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
0  Sales Executive                4        Single           5993        19479   

   NumCompaniesWorked Over18 OverTime  PercentSalaryHike  PerformanceRating  \
0                   8      Y      Yes                 11                  3   

   RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0                         1             80                 0   

   TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0                  8                      0                1               6   

   YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                   4                        0                     5

In [6]:
df.shape

(1470, 35)

In [7]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [0]:
#rearranging the columns with the target variable as the last column

col_sequence = ['Age', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'Attrition']

In [0]:
df = df.reindex(columns=col_sequence)

In [10]:
df.head()

Age     BusinessTravel  DailyRate              Department  \
0   41      Travel_Rarely       1102                   Sales   
1   49  Travel_Frequently        279  Research & Development   
2   37      Travel_Rarely       1373  Research & Development   
3   33  Travel_Frequently       1392  Research & Development   
4   27      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager Attrition  
0                        0                     5       Yes  
1                        1                     7        No  
2                        0                     0       Yes  
3                        3                     0        No  
4                        2                     2        No

# Initial EDA and Pre-Processing:

In [11]:
#checking for NA values

df.isnull().sum()

Age                         0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithC

let us look into the variables to determine which are numerical and which are categorical

In [12]:
for i in col_sequence:
  if (df[i].value_counts()).agg('count') <= 20:
    print (df[i].value_counts(dropna=False),'\n')

Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64 

Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64 

3    572
4    398
2    282
1    170
5     48
Name: Education, dtype: int64 

Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64 

1    1470
Name: EmployeeCount, dtype: int64 

3    453
4    446
2    287
1    284
Name: EnvironmentSatisfaction, dtype: int64 

Male      882
Female    588
Name: Gender, dtype: int64 

3    868
2    375
4    144
1     83
Name: JobInvolvement, dtype: int64 

1    543
2    534
3    218
4    106
5     69
Name: JobLevel, dtype: int64 

Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manag

**three variables stand out above: Over18, StandardHours, and EmployeeCount. All 3 have same value for all rows, hence these variables are useless. So we drop them.
Below we also see that the variable EmployeeNumber has all unique values (possibly it is an identification field). We drop this as well.**

**2nd point of note is the number of records having "Yes" in the Attrition variable - 237 out of 1470 (16%). Thus our dataset is imbalanced and we need to handle it accordingly..**

In [13]:
for i in col_sequence:
  if (df[i].value_counts()).agg('count') > 20:
    print (i,':',(df[i].value_counts()).agg('count'),'\n')

Age : 43 

DailyRate : 886 

DistanceFromHome : 29 

EmployeeNumber : 1470 

HourlyRate : 71 

MonthlyIncome : 1349 

MonthlyRate : 1427 

TotalWorkingYears : 40 

YearsAtCompany : 37 



In [0]:
df.EmployeeNumber.value_counts(dropna=False)

In [0]:
df = df.drop(['EmployeeCount', 'EmployeeNumber', 'Over18', 'StandardHours'], axis=1)

In [17]:
df.head()

Age     BusinessTravel  DailyRate              Department  \
0   41      Travel_Rarely       1102                   Sales   
1   49  Travel_Frequently        279  Research & Development   
2   37      Travel_Rarely       1373  Research & Development   
3   33  Travel_Frequently       1392  Research & Development   
4   27      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   
2                 2          2          Other                        4   
3                 3          4  Life Sciences                        4   
4                 2          1        Medical                        1   

   Gender  HourlyRate  JobInvolvement  JobLevel                JobRole  \
0  Female          94               3         2        Sales Executive   
1    Male          61               2         2     Research Scientist   
2    Male          92               2         1  Laboratory Technician   
3  Female          56               3         1     Research Scientist   
4    Male          40               3         1  Laboratory Technician   

   JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
0                4        Single           5993        19479   
1                2       Married           5130        24907   
2                3        Single           2090         2396   
3                3       Married           2909        23159   
4                2       Married           3468        16632   

   NumCompaniesWorked OverTime  PercentSalaryHike  PerformanceRating  \
0                   8      Yes                 11                  3   
1                   1       No                 23                  4   
2                   6      Yes                 15                  3   
3                   1      Yes                 11                  3   
4                   9       No                 12                  3   

   RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
0                         1                 0                  8   
1                         4                 1                 10   
2                         2                 0                  7   
3                         3                 0                  8   
4                         4                 1                  6   

   TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                      0                1               6                   4   
1                      3                3              10                   7   
2                      3                3               0                   0   
3                      3                3               8                   7   
4                      3                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager Attrition  
0                        0                     5       Yes  
1                        1                     7        No  
2                        0                     0       Yes  
3                        3                     0        No  
4                        2                     2        No

**From data dictionary:**

Education:
  1 'Below College'
	2 'College'
	3 'Bachelor'
	4 'Master'
	5 'Doctor'
	
EnvironmentSatisfaction:
  1 'Low'
	2 'Medium'
	3 'High'
	4 'Very High'
	
JobInvolvement:
  1 'Low'
	2 'Medium'
	3 'High'
	4 'Very High'
	
JobSatisfaction:
  1 'Low'
	2 'Medium'
	3 'High'
	4 'Very High'
	
PerformanceRating:
  1 'Low'
	2 'Good'
	3 'Excellent'
	4 'Outstanding'
	
RelationshipSatisfaction:
  1 'Low'
	2 'Medium'
	3 'High'
	4 'Very High'
	
WorkLifeBalance:
  1 'Bad'
	2 'Good'
	3 'Better'
	4 'Best'
  
  **So these are ordinal categorical variables and need to be treated separately from numerical variables.**

**Lets create two separate datasets, for categorical and numerical variables**

In [0]:
df.columns

Index(['Age', 'BusinessTravel', 'DailyRate', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'Gender',
       'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobRole',
       'JobSatisfaction', 'MaritalStatus', 'MonthlyIncome', 'MonthlyRate',
       'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'Attrition'],
      dtype='object')

In [0]:
df.head(2)

Age     BusinessTravel  DailyRate              Department  \
0   41      Travel_Rarely       1102                   Sales   
1   49  Travel_Frequently        279  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   

   Gender  HourlyRate  JobInvolvement  JobLevel             JobRole  \
0  Female          94               3         2     Sales Executive   
1    Male          61               2         2  Research Scientist   

   JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
0                4        Single           5993        19479   
1                2       Married           5130        24907   

   NumCompaniesWorked OverTime  PercentSalaryHike  PerformanceRating  \
0                   8      Yes                 11                  3   
1                   1       No                 23                  4   

   RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
0                         1                 0                  8   
1                         4                 1                 10   

   TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                      0                1               6                   4   
1                      3                3              10                   7   

   YearsSinceLastPromotion  YearsWithCurrManager Attrition  
0                        0                     5       Yes  
1                        1                     7        No

In [0]:
df_num = df.loc[:,['Age', 'DailyRate', 'DistanceFromHome', 'HourlyRate', 'MonthlyIncome', 'MonthlyRate',
       'NumCompaniesWorked', 'PercentSalaryHike', 'TotalWorkingYears', 'TrainingTimesLastYear', 
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']]

In [0]:
df_cat = df.loc[:,['BusinessTravel', 'Department', 'Education', 'EducationField', 'EnvironmentSatisfaction', 'Gender',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction', 'MaritalStatus', 'OverTime', 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'WorkLifeBalance', 'Attrition']]

In [42]:
df_num.head(2)

Age  DailyRate  DistanceFromHome  HourlyRate  MonthlyIncome  MonthlyRate  \
0   41       1102                 1          94           5993        19479   
1   49        279                 8          61           5130        24907   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  \
0                   8                 11                  8   
1                   1                 23                 10   

   TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                      0               6                   4   
1                      3              10                   7   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7

In [21]:
df_cat.head(2)

BusinessTravel              Department  Education EducationField  \
0      Travel_Rarely                   Sales          2  Life Sciences   
1  Travel_Frequently  Research & Development          1  Life Sciences   

   EnvironmentSatisfaction  Gender  JobInvolvement  JobLevel  \
0                        2  Female               3         2   
1                        3    Male               2         2   

              JobRole  JobSatisfaction MaritalStatus OverTime  \
0     Sales Executive                4        Single      Yes   
1  Research Scientist                2       Married       No   

   PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0                  3                         1                 0   
1                  4                         4                 1   

   WorkLifeBalance Attrition  
0                1       Yes  
1                3        No

# Categorical Variables:

**let us encode the categorical variables**

In [29]:
df_cat.BusinessTravel.value_counts()

Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64

**This can be encoded as an ordinal variable: non_travel as 0, travel_rarely as 1, travel_frequently as 2**

In [0]:
df_cat['BusinessTravel'].replace(to_replace=['Travel_Rarely', 'Travel_Frequently', 'Non-Travel'], value=[1,2,0], inplace=True)

In [31]:
df_cat.BusinessTravel.value_counts()

1    1043
2     277
0     150
Name: BusinessTravel, dtype: int64

In [33]:
cat_cols = ['Department', 'EducationField',
        'Gender', 'JobRole', 'MaritalStatus', 'OverTime', 'Attrition']
df_cat.loc[:,cat_cols].head()

Department EducationField  Gender                JobRole  \
0                   Sales  Life Sciences  Female        Sales Executive   
1  Research & Development  Life Sciences    Male     Research Scientist   
2  Research & Development          Other    Male  Laboratory Technician   
3  Research & Development  Life Sciences  Female     Research Scientist   
4  Research & Development        Medical    Male  Laboratory Technician   

  MaritalStatus OverTime Attrition  
0        Single      Yes       Yes  
1       Married       No        No  
2        Single      Yes       Yes  
3       Married      Yes        No  
4       Married       No        No

In [34]:
df_cat.Department.value_counts()

Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64

In [35]:
df_cat.EducationField.value_counts()

Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64

In [37]:
df_cat.JobRole.value_counts()

Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: JobRole, dtype: int64

In [38]:
df_cat.MaritalStatus.value_counts()

Married     673
Single      470
Divorced    327
Name: MaritalStatus, dtype: int64

In [39]:
df_cat.OverTime.value_counts()

No     1054
Yes     416
Name: OverTime, dtype: int64

In [40]:
df_cat.Gender.value_counts()

Male      882
Female    588
Name: Gender, dtype: int64

**Gender and OverTime can be label-encoded as they have just two labels.**

In [0]:
from sklearn.preprocessing import LabelEncoder

In [0]:
le = LabelEncoder()

In [0]:
df_cat_enc = df_cat.loc[:,['Gender', 'OverTime']].apply(le.fit_transform)

In [47]:
df_cat_enc.head()

Gender  OverTime
0       0         1
1       1         0
2       1         1
3       0         1
4       1         0

In [0]:
df_cat_oh = pd.get_dummies(df_cat, columns=['Department', 'EducationField', 'JobRole', 'MaritalStatus'], drop_first=True)

In [51]:
df_cat_oh.head()

BusinessTravel  Education  EnvironmentSatisfaction  Gender  JobInvolvement  \
0               1          2                        2  Female               3   
1               2          1                        3    Male               2   
2               1          2                        4    Male               2   
3               2          4                        4  Female               3   
4               1          1                        1    Male               3   

   JobLevel  JobSatisfaction OverTime  PerformanceRating  \
0         2                4      Yes                  3   
1         2                2       No                  4   
2         1                3      Yes                  3   
3         1                3      Yes                  3   
4         1                2       No                  3   

   RelationshipSatisfaction  StockOptionLevel  WorkLifeBalance Attrition  \
0                         1                 0                1       Yes   
1                         4                 1                3        No   
2                         2                 0                3       Yes   
3                         3                 0                3        No   
4                         4                 1                3        No   

   Department_Research & Development  Department_Sales  \
0                                  0                 1   
1                                  1                 0   
2                                  1                 0   
3                                  1                 0   
4                                  1                 0   

   EducationField_Life Sciences  EducationField_Marketing  \
0                             1                         0   
1                             1                         0   
2                             0                         0   
3                             1                         0   
4                             0                         0   

   EducationField_Medical  EducationField_Other  \
0                       0                     0   
1                       0                     0   
2                       0                     1   
3                       0                     0   
4                       1                     0   

   EducationField_Technical Degree  JobRole_Human Resources  \
0                                0                        0   
1                                0                        0   
2                                0                        0   
3                                0                        0   
4                                0                        0   

   JobRole_Laboratory Technician  JobRole_Manager  \
0                              0                0   
1                              0                0   
2                              1                0   
3                              0                0   
4                              1                0   

   JobRole_Manufacturing Director  JobRole_Research Director  \
0                               0                          0   
1                               0                          0   
2                               0                          0   
3                               0                          0   
4                               0                          0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                           0                        1   
1                           1                        0   
2                           0                        0   
3                           1                        0   
4                           0                        0   

   JobRole_Sales Representative  MaritalStatus_Married  MaritalStatus_Single  
0                             0                      0                     1  
1                             0                      1                     0  
2     

In [53]:
df_cat_oh.columns

Index(['BusinessTravel', 'Education', 'EnvironmentSatisfaction', 'Gender',
       'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'OverTime',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'WorkLifeBalance', 'Attrition', 'Department_Research & Development',
       'Department_Sales', 'EducationField_Life Sciences',
       'EducationField_Marketing', 'EducationField_Medical',
       'EducationField_Other', 'EducationField_Technical Degree',
       'JobRole_Human Resources', 'JobRole_Laboratory Technician',
       'JobRole_Manager', 'JobRole_Manufacturing Director',
       'JobRole_Research Director', 'JobRole_Research Scientist',
       'JobRole_Sales Executive', 'JobRole_Sales Representative',
       'MaritalStatus_Married', 'MaritalStatus_Single'],
      dtype='object')

In [0]:
df_cat_oh.drop(['BusinessTravel', 'Education', 'EnvironmentSatisfaction', 'Gender',
       'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'OverTime',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'WorkLifeBalance', 'Attrition'], axis=1, inplace=True)

In [55]:
df_cat_oh.head(1)

Department_Research & Development  Department_Sales  \
0                                  0                 1   

   EducationField_Life Sciences  EducationField_Marketing  \
0                             1                         0   

   EducationField_Medical  EducationField_Other  \
0                       0                     0   

   EducationField_Technical Degree  JobRole_Human Resources  \
0                                0                        0   

   JobRole_Laboratory Technician  JobRole_Manager  \
0                              0                0   

   JobRole_Manufacturing Director  JobRole_Research Director  \
0                               0                          0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                           0                        1   

   JobRole_Sales Representative  MaritalStatus_Married  MaritalStatus_Single  
0                             0                      0                     1

In [56]:
df_cat.columns

Index(['BusinessTravel', 'Department', 'Education', 'EducationField',
       'EnvironmentSatisfaction', 'Gender', 'JobInvolvement', 'JobLevel',
       'JobRole', 'JobSatisfaction', 'MaritalStatus', 'OverTime',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'WorkLifeBalance', 'Attrition'],
      dtype='object')

In [0]:
df_cat.drop(['Department', 'EducationField', 'JobRole', 'MaritalStatus', 'Gender', 'OverTime'], axis=1, inplace=True)

In [59]:
df_cat.head(1)

BusinessTravel  Education  EnvironmentSatisfaction  JobInvolvement  \
0               1          2                        2               3   

   JobLevel  JobSatisfaction  PerformanceRating  RelationshipSatisfaction  \
0         2                4                  3                         1   

   StockOptionLevel  WorkLifeBalance Attrition  
0                 0                1       Yes

In [65]:
df_cat.iloc[:,-1].head(1)

0    Yes
Name: Attrition, dtype: object

In [0]:
df_cat_final = pd.concat([df_cat.iloc[:,:-1], df_cat_enc, df_cat_oh, df_cat.iloc[:,-1]], axis=1)

In [72]:
df_cat_final.head(1)

BusinessTravel  Education  EnvironmentSatisfaction  JobInvolvement  \
0               1          2                        2               3   

   JobLevel  JobSatisfaction  PerformanceRating  RelationshipSatisfaction  \
0         2                4                  3                         1   

   StockOptionLevel  WorkLifeBalance  Gender  OverTime  \
0                 0                1       0         1   

   Department_Research & Development  Department_Sales  \
0                                  0                 1   

   EducationField_Life Sciences  EducationField_Marketing  \
0                             1                         0   

   EducationField_Medical  EducationField_Other  \
0                       0                     0   

   EducationField_Technical Degree  JobRole_Human Resources  \
0                                0                        0   

   JobRole_Laboratory Technician  JobRole_Manager  \
0                              0                0   

   JobRole_Manufacturing Director  JobRole_Research Director  \
0                               0                          0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                           0                        1   

   JobRole_Sales Representative  MaritalStatus_Married  MaritalStatus_Single  \
0                             0                      0                     1   

   Attrition  
0          1

In [0]:
df_cat_final['Attrition'].replace(to_replace=['Yes', 'No'], value=[1, 0], inplace=True)

In [73]:
df_cat_final.head(1)

BusinessTravel  Education  EnvironmentSatisfaction  JobInvolvement  \
0               1          2                        2               3   

   JobLevel  JobSatisfaction  PerformanceRating  RelationshipSatisfaction  \
0         2                4                  3                         1   

   StockOptionLevel  WorkLifeBalance  Gender  OverTime  \
0                 0                1       0         1   

   Department_Research & Development  Department_Sales  \
0                                  0                 1   

   EducationField_Life Sciences  EducationField_Marketing  \
0                             1                         0   

   EducationField_Medical  EducationField_Other  \
0                       0                     0   

   EducationField_Technical Degree  JobRole_Human Resources  \
0                                0                        0   

   JobRole_Laboratory Technician  JobRole_Manager  \
0                              0                0   

   JobRole_Manufacturing Director  JobRole_Research Director  \
0                               0                          0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                           0                        1   

   JobRole_Sales Representative  MaritalStatus_Married  MaritalStatus_Single  \
0                             0                      0                     1   

   Attrition  
0          1

# Numerical Variables:

In [75]:
df_num.head(1)

Age  DailyRate  DistanceFromHome  HourlyRate  MonthlyIncome  MonthlyRate  \
0   41       1102                 1          94           5993        19479   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  \
0                   8                 11                  8   

   TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                      0               6                   4   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5